src/server/frontend_xwayland/xwayland_server.cpp

```diff
@@ -355,19 +355,27 @@ int create_socket(struct sockaddr_un *addr, size_t path_size) {
 
     }
 
+auto const x11_lock_fmt = "/tmp/.X%d-lock";
+auto const x11_socket_fmt = "/tmp/.X11-unix/X%d";
+
 // TODO this can be written with more modern c++
 int create_lockfile(int xdisplay)
 {
     char lockfile[256];
 
-    snprintf(lockfile, sizeof lockfile, "/tmp/.X%d-lock", xdisplay);
+    snprintf(lockfile, sizeof lockfile, x11_lock_fmt, xdisplay);
 
     mir::Fd const fd{open(lockfile, O_WRONLY | O_CLOEXEC | O_CREAT | O_EXCL, 0444)};
     if (fd < 0)
         return EEXIST;
 
     int const size = dprintf(fd, "%10d\n", getpid());
-    if (size != 11)
+
+    // Check if anyone else has created the socket (even though we have the lockfile)
+    char x11_socket[256];
+    snprintf(x11_socket, sizeof x11_socket, x11_socket_fmt, xdisplay);
+
+    if (size != 11 || access(x11_socket, F_OK) == 0)
```

Comment:
Where did the 11 come from? I think the answer is `"%10d\n"` will end up being 11 bytes, which is non-obvious. It is especially non-obvious that this magic number would have to be updated whenever that format string changes. I don't see any need to verify a specific number of bytes were written, just that the write was successful. Perhaps this would be better:
```
bool const success_writing_to_lock_file = dprintf(fd, "%10d\n", getpid()) > 0;

...

if (success_writing_to_lock_file && access(x11_socket, F_OK) == 0)
```

Reply:
As we have no tests of this area of the system I was aiming to fix the logic in the least invasive way.

But sure, I can rework.